src/platform/MpiServer.c

```diff
@@ -0,0 +1,618 @@
+    {
+        OsConfigLogError(GetPlatformLog(), "Failed to parse client session from request body");
+        return HTTP_BAD_REQUEST;
+    }
+
+    if (NULL == (payload = JsonSerializeToString(request, g_payload)))
+    {
+        OsConfigLogError(GetPlatformLog(), "Failed to parse payload from request body");
+        return HTTP_BAD_REQUEST;
+    }
+
+    if (IsFullLoggingEnabled())
+    {
+        OsConfigLogInfo(GetPlatformLog(), "Received MpiSetDesired request, session '%s'", clientSession);
+    }
+
+    if (MPI_OK != mpiSetDesired((MPI_HANDLE)clientSession, (MPI_JSON_STRING)payload, strlen(payload)))
+    {
+        status = HTTP_BAD_REQUEST;
+    }
+
+    return status;
+}
+
+static HTTP_STATUS MpiGetReportedRequest(JSON_Object* request, char** response, int* responseSize, int(*mpiGetReported)(MPI_HANDLE, MPI_JSON_STRING*, int*))
+{
+    HTTP_STATUS status = HTTP_OK;
+    const char* clientSession = NULL;
+
+    if (NULL == request)
+    {
+        OsConfigLogError(GetPlatformLog(), "Invalid null request");
+        return HTTP_BAD_REQUEST;
+    }
+
+    if (NULL == (clientSession = JsonGetString(request, g_clientSession)))
+    {
+        OsConfigLogError(GetPlatformLog(), "Failed to parse client session from request body");
+        return HTTP_BAD_REQUEST;
+    }
+
+    if (IsFullLoggingEnabled())
+    {
+        OsConfigLogInfo(GetPlatformLog(), "Received MpiGetReported request, session '%s'", clientSession);
+    }
+
+    if (MPI_OK != mpiGetReported((MPI_HANDLE)clientSession, (MPI_JSON_STRING*)response, responseSize))
+    {
+        status = HTTP_BAD_REQUEST;
+    }
+
+    return status;
+}
+
+HTTP_STATUS MpiRequest(const char* uri, const char* requestBody, char** response, int* responseSize)
+{
+    HTTP_STATUS status = HTTP_OK;
+    JSON_Value* rootValue = NULL;
+    JSON_Object* request = NULL;
+
+    if (NULL == uri)
+    {
+        OsConfigLogError(GetPlatformLog(), "MpiRequest: called with invalid null URI");
+        return HTTP_BAD_REQUEST;
+    }
+
+    if (NULL == requestBody)
+    {
+        OsConfigLogError(GetPlatformLog(), "MpiRequest(%s): called with invalid null request body", uri);
+        return HTTP_BAD_REQUEST;
+    }
+
+    if (NULL == response)
+    {
+        OsConfigLogError(GetPlatformLog(), "MpiRequest(%s): called with invalid null response", uri);
+        return HTTP_BAD_REQUEST;
+    }
+
+    if (NULL == responseSize)
+    {
+        OsConfigLogError(GetPlatformLog(), "MpiRequest(%s): called with invalid null response size", uri);
+        return HTTP_BAD_REQUEST;
+    }
+
+    if (NULL == (rootValue = json_parse_string(requestBody)))
+    {
+        OsConfigLogError(GetPlatformLog(), "MpiRequest(%s): failed to parse request body", uri);
+        return HTTP_BAD_REQUEST;
+    }
+
+    if (NULL == (request = json_value_get_object(rootValue)))
+    {
+        OsConfigLogError(GetPlatformLog(), "MpiRequest(%s): failed to parse request object", uri);
+        return HTTP_BAD_REQUEST;
+    }
+
```

Comment:
Ouch. Meant to stop commenting on this issue with many returns from same function but this one is extreme, not ok. Use a status and if-else